Add redirects (#47)
rsksmart/devportal: tutorials/compile-smart-contracts/go.md

---
layout: rsk
title: Compile Smart Contracts to Go
---

> ###### (instructions valid for Ubuntu) 

Reviewer: ```suggestion
> These instructions are valid for Ubuntu but links to run in different operating systems are included 
```

Reviewer 2: we can add instructions for other operating systems, but i need to know specifically which other ones we want to include.

this is because i have to first figure out an alternative set of instructions for each OS we want to add (as i have only done this in ubuntu) and then try it out as well, so it's not a trivial change.

thanks

In order to interact with a Smart Contract in a Go application, the ABI ([application binary interface](https://solidity.readthedocs.io/en/latest/abi-spec.html)) of the contract must first be generated, and then compiled so that it can be imported in said application.

> # 1. Prerequisites

Reviewer: ```suggestion
## 1. Prerequisites
```

> ## 1.1 Solidity compiler

Reviewer: ```suggestion
### 1.1 Solidity compiler
```

> Also known as `solc`, can be installed through:

Reviewer: Via `snap` is not the only way to install `solc`. I would add this link https://solidity.readthedocs.io/en/v0.5.3/installing-solidity.html

Reviewer 2: added the link, thanks

```sh
sudo snap install solc --edge
```

> ## 1.2 `abigen` tool

Reviewer: ```suggestion
### 1.2 `abigen` tool
```

This is the tool used to generate the `.go` files with the contract ABI as input.

### 1.2.1 `protoc` tool (dependency)
Check [latest](https://github.com/protocolbuffers/protobuf/releases) stable version and replace `PROTOC_ZIP` accordingly.
```
PROTOC_ZIP=protoc-x.x.x-linux-x86_64.zip
curl -OL https://github.com/google/protobuf/releases/download/vx.x.x/$PROTOC_ZIP
sudo unzip -o $PROTOC_ZIP -d /usr/local bin/protoc
sudo unzip -o $PROTOC_ZIP -d /usr/local include/*
rm -f $PROTOC_ZIP
```

> 

Reviewer: I find suitable to add `protobuf` docs link too https://developers.google.com/protocol-buffers/docs/gotutorial

Reviewer 2: not sure i agree, we aren't doing anything with protocol buffers other than installing `protoc` as a dependency for `devtools`/`abigen`.

------

> Install `abigen` by doing:

Reviewer: This should go under the respective title in the correct order it must be executed

Reviewer 2: it's already in order, but i have added another title to make it more explicit.

```sh
go get -u github.com/ethereum/go-ethereum
cd $GOPATH/src/github.com/ethereum/go-ethereum/
make
make devtools
```

> ## 1.3 ABI file

Reviewer: ```suggestion
### 1.3 ABI file
```

Save the Smart Contract ABI to your local drive and keep the location at hand.

> If you do not have an ABI, you can generate one from the contract like so (replace the `.sol` file with your own):

Reviewer: ```suggestion
If you do not have an ABI, you can generate one from the contract replacing the `.sol` file with your own smart contract name:
```

```sh
solc --abi Contract.sol -o build
```
For this example, the resulting file will be `./build/Contract.abi`.

> # 2. Compilation

Reviewer: ```suggestion
## 2. Compilation
```

To convert the ABI into a Go file, execute the following command:
```sh
abigen --abi=./build/Contract.abi --pkg=mypackage --out=Contract.go
```
Replace the contract and package names with your own.

> ## 2.1 Example

Reviewer: ```suggestion
## Example
```
